cli/tools/registry/module_content.rs

```diff
@@ -102,4 +141,289 @@ impl ModuleContentProvider {
+    text_changes: &mut Vec<TextChange>,
+  ) -> Result<(), AnyError> {
+    static JSX_RUNTIME_RE: Lazy<regex::Regex> =
+      lazy_regex::lazy_regex!(r"(?i)^[\s*]*@jsxRuntime\s+(\S+)");
+    static JSX_FACTORY_RE: Lazy<regex::Regex> =
+      lazy_regex::lazy_regex!(r"(?i)^[\s*]*@jsxFactory\s+(\S+)");
+    static JSX_FRAGMENT_FACTORY_RE: Lazy<regex::Regex> =
+      lazy_regex::lazy_regex!(r"(?i)^[\s*]*@jsxFragmentFactory\s+(\S+)");
+
+    let start_pos = if parsed_source.program_ref().shebang().is_some() {
+      match text_info.text_str().find('\n') {
+        Some(index) => index + 1,
+        None => return Ok(()), // nothing in this file
+      }
+    } else {
+      0
+    };
+    let had_change = Cell::new(false);
+    let mut add_text_change = |new_text: String| {
+      had_change.set(true);
+      text_changes.push(TextChange {
+        range: start_pos..start_pos,
+        new_text,
+      })
+    };
+    let jsx_options =
+      self.resolve_jsx_options(specifier, text_info, diagnostic_reporter)?;
+    let leading_comments = parsed_source.get_leading_comments();
+    let leading_comments_has_re = |regex: &regex::Regex| {
+      leading_comments
+        .as_ref()
+        .map(|comments| {
+          comments.iter().any(|c| {
+            c.kind == deno_ast::swc::common::comments::CommentKind::Block
+              && regex.is_match(c.text.as_str())
+          })
+        })
+        .unwrap_or(false)
+    };
+    if !leading_comments_has_re(&JSX_RUNTIME_RE) {
+      add_text_change(format!(
+        "/** @jsxRuntime {} */",
+        jsx_options.jsx_runtime,
+      ));
+    }
+    if module_info.jsx_import_source.is_none() {
+      if let Some(import_source) = jsx_options.jsx_import_source {
+        add_text_change(format!("/** @jsxImportSource {} */", import_source));
+      }
+    }
+    if module_info.jsx_import_source_types.is_none() {
+      if let Some(import_source) = jsx_options.jsx_import_source_types {
+        add_text_change(format!(
+          "/** @jsxImportSourceTypes {} */",
+          import_source
+        ));
+      }
+    }
+    if !leading_comments_has_re(&JSX_FACTORY_RE) {
+      add_text_change(format!(
+        "/** @jsxFactory {} */",
+        jsx_options.jsx_factory,
+      ));
+    }
+    if !leading_comments_has_re(&JSX_FRAGMENT_FACTORY_RE) {
+      add_text_change(format!(
+        "/** @jsxFragmentFactory {} */",
+        jsx_options.jsx_fragment_factory,
+      ));
+    }
+    // add a newline at the end
+    if had_change.get() {
+      let file_text = text_info.text_str();
+      let is_crlf = file_text
+        .find('\n')
+        .and_then(|index| {
+          let index_before = index.checked_sub(1)?;
+          let char_before = file_text.get(index_before..index)?;
+          Some(matches!(char_before, "\r"))
+        })
+        .unwrap_or(false);
+      add_text_change(if is_crlf {
+        "\r\n".to_string()
+      } else {
+        "\n".to_string()
+      });
```

Comment:
I don't think we should not add newlines, to make line/col still mostly correct (because we do not source map).

Reply:
I initially didn’t have this, but the issue showed this. I’ll remove it